h/api/storage.py

```diff
@@ -44,6 +46,154 @@ def fetch_annotation(request, id):
     return elastic.Annotation.fetch(id)
 
 
+def _legacy_create_annotation_in_elasticsearch(request, data):
+    annotation = elastic.Annotation(data)
+    # FIXME: this should happen when indexing, not storing.
+    _prepare(request, annotation)
+    annotation.save()
+    return annotation
+
+
+def transform_links(document, claimant, created, updated):
+    """Return one or more document URI dicts for the given document data.
+
+    Returns a document URI dict for each of the URIs that the given document
+    data claims is equivalent to the claimant URI. This list will include at
+    least one "self-claim" document URI dict whose URI is the claimant URI
+    itself.
+
+    Each of the returned dicts contains all of the fields necessary to
+    initialize an h.api.models.DocumentURI object.
+
+    This function just returns new document URI dicts blindly, it does not
+    check whether equivalent DocumentURI objects already exist in the database.
+
+    :param document: the "document" sub-object that was POSTed to the API as
+        part of a new or updated annotation
+    :type document: dict
+
+    :param claimant: the URI that the browser was at when this annotation was
+        created (the top-level "uri" field of the annotation)
+    :type claimant: unicode
+
+    :param created: the time when the annotation was created
+    :type created: datetime.datetime
+
+    :param updated: the time when the annotation was last updated
+    :type updated: datetime.datetime
+
+    :returns: a list of one or more document URI dicts
+    :rtype: list of dicts
+
+    """
+    transformed = []
+    links = document.get('link', [])
+
+    for link in links:
+        # disregard self-claim urls as they're are being added separately
+        # later on.
+        if link.keys() == ['href'] and link['href'] == claimant:
+            continue
+
+        # disregard doi links as these are being added separately from the
+        # highwire and dc metadata later on.
+        if link.keys() == ['href'] and link['href'].startswith('doi:'):
+            continue
+
+        uri_ = link['href']
+        type = None
+
+        # highwire pdf (href, type=application/pdf)
+        if set(link.keys()) == set(['href', 'type']) and len(link.keys()) == 2:
```

Comment:
I know it's not your fault, but as you pointed out this statement is wrong.

To add some more context to what I wrote in Slack:

We should certainly check that the value of `type` is `application/pdf`. But after thinking more about it, we should also discard that link and generate it server-side from the highwire meta data (`document[highwire][pdf_url]`) to keep it consistent with how we handle the doi links.
This also has the added benefit of us getting way more `highwire-pdf` document uris, since we have more documents with this document meta data vs uri data.

Thoughs? (plus @nickstenning)